### PR TITLE
HTML API: Adjust code styling to Gutenberg's linter's preferences

### PR DESCRIPTION
**Pull request by dmsnell**
Trac: [#58918-ticket](https://core.trac.wordpress.org/ticket/58918#ticket)

In this patch we're adjusting the code style according to the rules that the linting process in Gutenberg requires.

There are only a couple code changes that should have no effect on the runtime:
 - a missing check to verify that only `UTF-8` is supported has been added (brought up because it was identified as an undefined variable)
 - a few `return false;` statements have been added to avoid having the linter complain that functions don't return a value despite indicating they return `bool`. the functions are stubs for coming support and currently `throw`, so the `return` statements are unreachable.

### CODE REVIEW

**Review by costdev — APPROVED** [review]
Thanks for the PR @dmsnell!

These changes look good to go! The same type of `@see` adjustments are needed elsewhere in Core too as they break the WordPress.org parser due to a lack of `{}`. We have a ticket for those but I don't believe the ones in this PR were covered by its patch. Great to see them in this PR (no pun intended).

File: src/wp-includes/html-api/class-wp-html-processor.php [hunk 1]
```diff
@@ -232,16 +232,16 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 	 * @return WP_HTML_Processor|null The created processor if successful, otherwise null.
 	 */
 	public static function createFragment( $html, $context = '<body>', $encoding = 'UTF-8' ) {
-		if ( '<body>' !== $context ) {
+		if ( '<body>' !== $context || 'UTF-8' !== $encoding ) {
```
Comment (dream-encode) [inline src/wp-includes/html-api/class-wp-html-processor.php:236]: @dmsnell The PR mentions that there should be no code changes.  Is this change intentional?

Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php:236]: good eye @dream-encode - I didn't mean to mislead there. this should be the single code change, and when I overlooked it it's because it was always intended to be the case to only support `UTF-8` and bail if any others are passed. all the docs omit this argument but it's important to be there as we continue to build support into the class.

hopefully there should be no behavioral change, unless someone was intentionally trying to run this in an unsupported mode (which should not be the case as this only merged into `trunk` a few days ago).

Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-processor.php:236]: I've updated the description to better explain

File: src/wp-includes/html-api/class-wp-html-open-elements.php
Comment (peterwilsoncc) [inline src/wp-includes/html-api/class-wp-html-open-elements.php:1]: I'm unclear, why are there so many unused variables and unimplemented functions in this file?

File: src/wp-includes/html-api/class-wp-html-active-formatting-elements.php [hunk 2]
```diff
@@ -149,7 +148,7 @@ public function remove_node( $token ) {
 	 *     > EM -> STRONG -> A ->
 	 *
 	 * To start with the most-recently added element and walk towards the top,
-	 * @see WP_HTML_Active_Formatting_Elements::walk_up
+	 * {@see WP_HTML_Active_Formatting_Elements::walk_up}.
```
Comment (peterwilsoncc) [inline src/wp-includes/html-api/class-wp-html-active-formatting-elements.php]:
```suggestion
	 * see WP_HTML_Active_Formatting_Elements::walk_up().
```

For functions methods, the developer.wordpress.org docs parser will pick up links if brackets are included, `{@see` is only needed for hooks. See `get_posts()` for an example:

https://github.com/WordPress/wordpress-develop/blob/444bdf49bc38883f603a584bcc119aedf98257b8/src/wp-includes/post.php#L2392

:1234: Applies throughout.

File: src/wp-includes/html-api/class-wp-html-open-elements.php
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-open-elements.php:1]: Hi @peterwilsoncc! thanks for taking a look.

the reason for this odd-seeming code is mostly to build out the interfaces we know we're going to need, but provide a way to implement HTML compliance without dumping a single massive patch.

we've chosen to build this in stages to give enough opportunity to review each stage and focus on testing each step and the implications of each step. in accordance with the philosophy started with the Tag Processor, we want to always avoid breaking, but we're find aborting. that is, this file and others have the scaffolding in place where we'll need to expand support when we add support for the kinds of HTML tags that make those functions relevant. today, it's not possible to get into a spot where those functions would be called, and where the variables are unused, it''s not possible to get into a spot where they would be meaningful.

they are there because that's what the function interface needs to be, and a large part of this is building a clear correspondence between the implementation code and the HTML specification; this in order to make it easier to expand and maintain and debug. most of these function names and interfaces are lifted directly from the spec.

as we continue to expand support in the HTML Processor, which we'll do at a pace that gives the ability to review and test and vet each additional bit of support, this will gradually fill up until those variables are all used and all the functions are fully implemented. the alternative is a giant code dump and a "let's hope this works and please trust that it does because there's no way we can examine it." 😄 

for more insight into where this structure leads you can look at https://github.com/dmsnell/wordpress-develop/pull/6 and https://github.com/adamziel/wordpress-develop/pull/2.

For this specific patch we're mostly trying to clear up the obstacles that are preventing the code that has been merged into Core from being backported into Gutenberg.

File: src/wp-includes/html-api/class-wp-html-active-formatting-elements.php (hunk 2 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-active-formatting-elements.php]: Are you suggesting we remove the brackets? If so, I have a couple questions:
 - although the docs parser might pick them up, what constitutes a link and how can we tell before the docs are published and updated if we did it right?
 - IDE support catches the `{@see …}` syntax as do all normal docblock parsing code. This is lost when removing the brackets. Are you saying you prefer removing the brackets and losing that support?
 - Are there any particular benefits to removing the brackets and the standard docblock tagging?

Comment (peterwilsoncc) [reply src/wp-includes/html-api/class-wp-html-active-formatting-elements.php]: > * lthough the docs parser might pick them up, what constitutes a link and how can we tell before the docs are published and updated if we did it right?

* For linking to hooks `{@see 'some_hook_name'}` constitutes a link (the quotes are required)
* For linking to functions and methods `function()` and `Class::method()` constitute links (the brackets are required)

You can validate that you've done it correctly but running [the docs parser](https://github.com/WordPress/phpdoc-parser) locally. Rather than generating the full docs, it's best to specify the file when running the command to avoid waiting for the full docs to be generated `wp parser create /path/to/file.php --user=<id|login>`

> Are you suggesting we remove the brackets?

Yeah, I'm suggesting removing the braces, and adding the brackets to the method names -- reviewing the WP Code it seems to be the standard. I'm basing this on a ticket opened by a meta team member after I was similarly unclear, see [WP#45445](https://core.trac.wordpress.org/ticket/45445). The key quote being:

> The documentation standards [mention](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#other-descriptions) the use of @see for referencing hooks, but could stand to be more explicit about not needing to be used for functions, classes, or class methods.

File: src/wp-includes/html-api/class-wp-html-open-elements.php
Comment (peterwilsoncc) [reply src/wp-includes/html-api/class-wp-html-open-elements.php:1]: I understand that it's being built up gradually, and agree it makes sense to, but I don't think it's a good idea to include the unused variables and unimplemented methods until they're required.

Including them locks WP into specifics that may need to change as the implementation is built out. I know this has been speced out quite thoroughly but including the unimplemented skeleton now makes promises to extenders that may not be kept.

File: src/wp-includes/html-api/class-wp-html-active-formatting-elements.php (hunk 2 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-active-formatting-elements.php]: > running [the docs parser](https://github.com/WordPress/phpdoc-parser) locally

thank you! I've asked before how I can preview the docs and this is the first time someone has indicated how. I'll poke around with that. seems like it needs to run on a copy of WordPress as a plugin.

> I'm suggesting removing the braces, and adding the brackets to the method names 

looks like I was confused on the distinction of braces and brackets and parenthesis.

I'm really bummed about removing syntax that works and breaks IDE integrations for that hides the links implicitly. do you know there's a preference not to use the standardized docblock syntax?

if I'm required to remove the curly brackets I'm more inclined to awkwardly rewrite the docblock so it can use a `@see` tag on its own. I'd be sad if people are prevented from navigating the code because this specific formalization isn't required in order for the docs to display correct.

any thoughts on this? is there wiggle-room to cater towards developers who are reading the code?

Comment (costdev) [reply src/wp-includes/html-api/class-wp-html-active-formatting-elements.php]: I thought there was nothing in the inline documentation standards that indicates "the right way" to do this, but I did see something in [this section](https://developer.wordpress.org/coding-standards/inline-documentation-standards/php/#phpdoc-tags):

> `@see` | elementname | References another function/method/class the function/method relies on. **Should only be used inline for “linking” hooks.**
(my emphasis)

There's certainly a difference between "should" and "must", and Core has instances of inline `@see` referencing functions. I'm not sure there's a solid answer to this one yet. We could probably do with some clarification in the standards.

I do see the very valid point about IDE support. I wonder if there's a reason why we _shouldn't_ use `{@see function_name()}`, other than it being unnecessary for the docs parser.

FYI: I'd opened [a ticket](https://core.trac.wordpress.org/ticket/58858) recently where I identified some `@see` tags broken on the parser. That ticket, and this PR could help standardize a consistent format regardless of what's being referenced, saving some brainpower.

Comment (peterwilsoncc) [reply src/wp-includes/html-api/class-wp-html-active-formatting-elements.php]: > I'm really bummed about removing syntax that works and breaks IDE integrations for that hides the links implicitly. do you know there's a preference not to use the standardized docblock syntax?

I'm really basing this on none of the existing code in core using `@see` for functions/methods (some external libraries do) and the ticket I linked to above.

FWIW, I think your point about supporting IDEs is a solid argument for starting a discussion on make/core. Apparently I don't have the right extension installed in VSCode for `@see` linking to work but it makes sense that IDEs would use it.

> I'll poke around with that. seems like it needs to run on a copy of WordPress as a plugin.

I recall it being a little fiddly to get set up but worth it for testing potential edge cases.

File: src/wp-includes/html-api/class-wp-html-open-elements.php
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-open-elements.php:1]: > Including them locks WP into specifics that may need to change as the implementation is built out. I know this has been spec'd out quite thoroughly but including the unimplemented skeleton now makes promises to extenders that may not be kept.

hm. this is exactly the point of adding them, to guard against updates that might overlook it or against people starting to rely on having those functions with one argument when they'll come back with more. we can provide dummy values, but I don't see that adding much.

also I think _within weeks_ we might have a number of these filled out, so they may not last very long as unused. in fact, I'm hoping to have the use-case ready for the lightbox work in time for 6.4, and I think that will pull in lots of this in the process.

File: src/wp-includes/html-api/class-wp-html-active-formatting-elements.php (hunk 2 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-active-formatting-elements.php]: Are we good to go with the syntax as written and continue considering the special syntax for WordPress in follow-up work that focuses on that PHPDoc tag and whether to allow the inline form?

Comment (peterwilsoncc) [reply src/wp-includes/html-api/class-wp-html-active-formatting-elements.php]: I'd rather go with the existing standard in WP core, ie `see WP_HTML_Active_Formatting_Elements::walk_up()` and the equivalent elsewhere. A make/core post can be used to discuss whether we change the standard.

If this PR uses a special case, all that is going to happen is someone will do a follow up commit fixing the coding standards for consistency.

Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-active-formatting-elements.php]: I will update to the existing pattern, though I thought we established it's not truly a standard; more like nothing else in Core has used this legitimate pattern.

Will try and post this quickly so it doesn't hold up the patch. Thanks @peterwilsoncc

Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-active-formatting-elements.php]: updated in cdd785cbbf

the language is awkward with `see ` coming out of nowhere, so I had to smudge a few places and use a normal `@see` tag to avoid making it confusing, and we've lost the ability to lean on tools provided by the IDE, but the inline syntax is gone.

File: src/wp-includes/html-api/class-wp-html-token.php [hunk 3]
```diff
@@ -36,10 +36,12 @@ class WP_HTML_Token {
 	/**
 	 * Name of node; lowercase names such as "marker" are not HTML elements.
 	 *
-	 * For HTML elements/tags this value should come from {@see WP_HTML_Processor::get_tag}.
+	 * For HTML elements/tags this value should come from `get_tag`
```
Comment (SergeyBiryukov) [inline src/wp-includes/html-api/class-wp-html-token.php]:
```suggestion
	 * For HTML elements/tags this value should come from WP_HTML_Processor::get_tag().

Comment (ockham) [reply src/wp-includes/html-api/class-wp-html-token.php]: This makes sense to me, as the method we're referencing here isn't part of this class.

@dmsnell I guess you'll also want to keep [the `@see WP_HTML_Processor::get_tag()` on R43](https://github.com/WordPress/wordpress-develop/pull/4918/files#diff-a2a58890c64ceffde44307c625eefd5fe11162ac07b0e72f2d6792df2c90219bR43) for better IDE integration, and per precedent elsewhere in this patch?

**Review by ockham — APPROVED** [review]
LGTM; [Sergey's suggestion](https://github.com/WordPress/wordpress-develop/pull/4918/files#r1285061392) makes sense to me, though.

File: src/wp-includes/html-api/class-wp-html-token.php (hunk 3 above)
Comment (ockham) [reply src/wp-includes/html-api/class-wp-html-token.php]: I'll take the liberty of applying this suggestion and will then commit this to svn in order to unblock work on the HTML API 😊

**Comment by ockham** [comment]
Thanks all! Committed to Core in https://core.trac.wordpress.org/changeset/56363/.

File: src/wp-includes/html-api/class-wp-html-token.php (hunk 3 above)
Comment (dmsnell) [reply src/wp-includes/html-api/class-wp-html-token.php]: if you trace the changes you will note that I reworded this to avoid what felt like an awkward sentence without the inline syntax to delimit it. I added a full `@see` tag below, so I guess things are duplicated now (which is why I removed the class reference, since it sits in full in lines 43 and 44 immediately below it).

https://github.com/WordPress/wordpress-develop/pull/4918/files#diff-a2a58890c64ceffde44307c625eefd5fe11162ac07b0e72f2d6792df2c90219bR43-R44

this is one of those cases where I felt like there wasn't a good option once we removed the PHPDoc syntax. I might try and update this further because now it's weird again with a duplicate reference right underneath this pseudo-reference.
